r2-core/src/main/java/com/linkedin/r2/transport/http/client/EvictingCircularBuffer.java

```diff
@@ -0,0 +1,245 @@
+          }
+        }
+        finally
+        {
+          thisLock.writeLock().unlock();
+        }
+      }
+      return get();
+    }
+    return callback;
+  }
+
+  /**
+   * @return the number of unique Callbacks this buffer can hold.
+   */
+  int getCapacity()
+  {
+    return _callbacks.size();
+  }
+
+  /**
+   * Resizes the circular buffer, deleting the contents in the process.
+   * This method should not be called frequently, ideally only as part of a startup lifecycle, as it does heavy locking
+   * to ensure all reads and writes are drained coinciding with the resizing of the buffer.
+   * @param capacity
+   */
+  void setCapacity(int capacity)
+  {
+    if (capacity < 1)
+    {
+      throw new IllegalArgumentException("capacity can't be less than 1");
+    }
+    // acquire write lock for all elements in the buffer to prevent reads while the buffer is re-created,
+    // taking care to store them in a temporary location for releasing afterward.
+    ArrayList<ReentrantReadWriteLock> tempLocks = new ArrayList<>();
+    _elementLocks.forEach(x ->
+    {
+      x.writeLock().lock();
+      tempLocks.add(x);
+    });
+    try
+    {
+      _callbacks.clear();
+      _ttlBuffer.clear();
+      _elementLocks.clear();
+      // populate ArrayList with nulls to prevent changes to underlying data structure size during writes,
+      // also needed to compute reader and writer position through calls to size()
+      _ttlBuffer.addAll(Collections.nCopies(capacity, null));
+      _callbacks.addAll(Collections.nCopies(capacity, null));
+      for(int i = 0; i <= capacity; i++)
+      {
+        _elementLocks.add(new ReentrantReadWriteLock());
+      }
+    }
+    finally
+    {
+      // these locks no longer exist in _elementLocks, but we need to release them in order to unblock
+      // pending reads.
+      tempLocks.forEach(x -> x.writeLock().unlock());
+    }
+  }
+
+  /**
+   * @return the currently configured TTL.
+   */
+  Duration getTtl()
+  {
+    return _ttl;
+  }
+
+  /**
+   * Sets the amount of time a Callback is eligible to be returned after it has been stored in the buffer.
+   * TTL is shared across all stored Callbacks for the sake of simplicity.
+   * @param ttl number value of amount of time
+   * @param ttlUnit unit of time for number value
+   */
+  void setTtl(int ttl, ChronoUnit ttlUnit)
+  {
+    if (ttl < 1)
+    {
+      throw new IllegalArgumentException("ttl can't be less than 1");
+    }
+    if (ttlUnit == null)
+    {
+      throw new IllegalArgumentException("ttlUnit can't be null.");
+    }
+    _ttl = Duration.of(ttl, ttlUnit);
+  }
+
+  private int getAndBumpWriterPosition()
+  {
+    return (_writerPosition.getAndUpdate(x -> (x + 1) % _callbacks.size()));
+  }
+
+  private int getAndBumpReaderPosition()
+  {
+    return (_readerPosition.getAndUpdate(x -> (x + 1) % _callbacks.size()));
+  }
+
+  private int size()
+  {
+    return (int) _callbacks.stream().filter(Objects::nonNull).count();
```

Comment:
If we can keep an atomic counter of the size, we don't need to go through the entire array each time we call size()

Reply:
size() method has been removed after implementing while-loop-based iteration on the get() call.